packages/xo-server-load-balancer/src/plan.js

```
@@ -777,11 +809,311 @@ export default class Plan {
+            sourceHost,
+            destinationHost,
+            vm,
+            hostsAverages,
+            vmAverages: vmsAverages[vm.id],
+          })
+        )
+      }
+    }
+    return promises
+  }
+
+  async _migrateOtherVms({ crowdedHost, hostsAverages, vmsAverages, idToHost, taggedHosts, memoryNeeded }) {
+    const promises = []
+
+    const candidateVms = sortBy(
+      filter(
+        Object.values(crowdedHost.vms),
+        vm => vm.xenTools && intersection(vm.tags, this._affinityTags).length === 0
+      ),
+      [vm => -vmsAverages[vm.id].memory] // try to migrate bigger VMs first to minimize the number of migrations
+    )
+    debugAffinity(`Candidate VMs to be moved away: ${candidateVms.map(vm => vm.name_label)}`)
+
+    for (const vm of candidateVms) {
+      // try to migrate vm
+
+      const vmAverages = vmsAverages[vm.id]
+
+      const destinationId = sortBy(
+        Object.keys(hostsAverages),
+        [hostId => -hostsAverages[hostId].memoryFree, hostId => hostsAverages[hostId].cpu] // try to migrate to hosts with the most free space first
+      ).find(hostId => {
+        if (hostId === crowdedHost.id) {
+          return false
+        }
+        const destinationAverages = hostsAverages[hostId]
+        return (
+          destinationAverages.cpu + vmAverages.cpu <= this._thresholds.cpu.critical &&
+          destinationAverages.memoryFree - vmAverages.memory >= this._thresholds.memoryFree.critical
+        )
+      })
+
+      if (destinationId === undefined) {
+        // no host can accept this VM, let's try another one
+        debug(`Cannot migrate VM (${vm.id}) to any Host.`)
+        continue
+      }
+
+      // destination found, now migrate and update tags & averages
+      promises.push(
+        this._migrateVmAndUpdateInfos({
+          destination: idToHost[destinationId],
+          source: idToHost[crowdedHost.id],
+          sourceHost: crowdedHost,
+          destinationHost: taggedHosts[destinationId],
```

From what I see, `taggedHosts` is structured as `{ hosts: Host[], tagCount: { [tag: string]: number } }`, so indexing it with `destinationId` would give `undefined`, right?
Wouldn’t that throw when we later try to read properties on `destinationHost` (or cause silent failures)?
Did I misread the shape of `taggedHosts`, or should we probably do something like `taggedHosts.hosts.find(h => h.id === destinationId)` (or use the existing `idToHost` map we already have in scope)?